Eliminate string copying when formatting dates and times

PHILO-HE opened the pull request:
DateTimeFormatter::format API used to create intermediate strings resulting in
unnecessary data copying.

The new API allows the caller to check the max size of the formatted string,
pre-allocate that much memory, then call 'format' to write directly to
pre-allocated buffer.

This PR also includes a simple benchmark which shows ~2x acceleration with
this optimization.

See https://github.com/facebookincubator/velox/pull/7555/files#r1400271919

PHILO-HE commented:
cc @rui-mo

rui-mo commented on velox/functions/lib/DateTimeFormatter.cpp:
```diff
@@ -927,9 +928,113 @@ void parseFromPattern(
+// buffer.
+int32_t DateTimeFormatter::evaluateMaxResultSize(
+    const date::time_zone* timezone) const {
+  int32_t size = 0;
```
Maybe use uint32_t instead because size cannot be negative.

rui-mo commented on velox/functions/lib/DateTimeFormatter.cpp:
```diff
@@ -927,9 +928,113 @@ void parseFromPattern(
+int32_t DateTimeFormatter::evaluateMaxResultSize(
+    const date::time_zone* timezone) const {
+  int32_t size = 0;
+  for (auto& token : tokens_) {
```
const auto&

rui-mo commented on velox/functions/lib/DateTimeFormatter.cpp:
```diff
@@ -927,9 +928,113 @@ void parseFromPattern(
+        break;
+      case DateTimeFormatSpecifier::DAY_OF_WEEK_0_BASED:
+      case DateTimeFormatSpecifier::DAY_OF_WEEK_1_BASED:
+        size += std::max((int)token.pattern.minRepresentDigits, 1);
```
Seems same as `case DateTimeFormatSpecifier::CENTURY_OF_ERA:`. Maybe combine them.

rui-mo commented on velox/functions/lib/DateTimeFormatter.cpp:
```diff
@@ -927,9 +928,113 @@ void parseFromPattern(
+        size += std::max((int)token.pattern.minRepresentDigits, 1);
+        break;
+      case DateTimeFormatSpecifier::DAY_OF_WEEK_TEXT:
+        if (token.pattern.minRepresentDigits <= 3) {
```
Maybe simplify as below.
> size = token.pattern.minRepresentDigits <= 3 ? size + 3 : size +9;

rui-mo commented on velox/functions/lib/DateTimeFormatter.cpp:
```diff
@@ -927,9 +928,113 @@ void parseFromPattern(
+      case DateTimeFormatSpecifier::CLOCK_HOUR_OF_DAY:
+      case DateTimeFormatSpecifier::MINUTE_OF_HOUR:
+      case DateTimeFormatSpecifier::SECOND_OF_MINUTE:
+        size += std::max((int)token.pattern.minRepresentDigits, 2);
```
Same with `case DateTimeFormatSpecifier::DAY_OF_MONTH:` and `case DateTimeFormatSpecifier::MONTH_OF_YEAR:`.

rui-mo commented on velox/functions/lib/DateTimeFormatter.cpp:
```diff
@@ -927,9 +928,113 @@ void parseFromPattern(
+        break;
+      case DateTimeFormatSpecifier::TIMEZONE_OFFSET_ID:
+      case DateTimeFormatSpecifier::WEEK_YEAR:
+      case DateTimeFormatSpecifier::WEEK_OF_WEEK_YEAR:
```
Why are they added? Seems they will go to default behavior.

rui-mo commented on velox/functions/lib/DateTimeFormatter.cpp:
```diff
@@ -927,9 +928,113 @@ void parseFromPattern(
+            token.pattern.specifier);
+    }
+  }
+  return size;
```
Why not return in each branch, and remove these `break`?

PHILO-HE commented:
Hi @mbasmanova, could you take a review?

mbasmanova commented on velox/functions/prestosql/DateTimeFunctions.h:
```diff
@@ -1166,12 +1166,12 @@ struct DateFormatFunction : public TimestampWithTimezoneSupport<T> {
-    auto formattedResult = mysqlDateTime_->format(timestamp, sessionTimeZone_);
-    auto resultSize = formattedResult.size();
+    const auto maxSize =
+        mysqlDateTime_->evaluateMaxResultSize(sessionTimeZone_);
```
sessionTimeZone_ doesn't change, hence, we can evaluate this once, no?

mbasmanova commented on velox/functions/lib/DateTimeFormatter.cpp:
```diff
@@ -927,9 +928,92 @@ void parseFromPattern(
 
 } // namespace
 
+// Evaluates the max size of datetime format result to pre-allocate result
```
This comment belongs to .h file.

Returns max size of the formatted string. Can be used to preallocate memory before calling format() to avoid extra copy.

mbasmanova commented on velox/functions/lib/DateTimeFormatter.cpp:
```diff
@@ -927,9 +928,92 @@ void parseFromPattern(
+      case DateTimeFormatSpecifier::DAY_OF_WEEK_TEXT:
+      case DateTimeFormatSpecifier::MONTH_OF_YEAR_TEXT:
+        // 9 is the max size of elements in weekdaysFull or monthsFull.
+        size = token.pattern.minRepresentDigits <= 3 ? size + 3 : size + 9;
```
nit: use size += to avoid confusing the reader

mbasmanova commented on velox/functions/lib/DateTimeFormatter.cpp:
```diff
@@ -927,9 +928,92 @@ void parseFromPattern(
+        break;
+      case DateTimeFormatSpecifier::YEAR:
+        // Timestamp is in [-32767-01-01, 32767-12-31] range.
+        size = token.pattern.minRepresentDigits == 2
```
ditto

mbasmanova commented on velox/functions/lib/DateTimeFormatter.cpp:
```diff
@@ -927,9 +928,92 @@ void parseFromPattern(
+  return result;
+}
+
+// Result is pre-allocated according to evaluateMaxResultSize.
```
Move comments to .h

mbasmanova commented on velox/functions/lib/DateTimeFormatter.cpp:
```diff
@@ -927,9 +928,92 @@ void parseFromPattern(
+// Result is pre-allocated according to evaluateMaxResultSize.
+// Returns actual size.
+int32_t DateTimeFormatter::format(
+    const std::string_view& result,
```
this API is strange; std::string_view should be passed by value, but here I think we should use char* and make it last argument (input args first, output last).

mbasmanova commented on velox/functions/lib/DateTimeFormatter.cpp:
```diff
@@ -941,31 +1025,35 @@ std::string DateTimeFormatter::format(
 
-  std::string result;
+  char* resultBuffer = const_cast<char*>(result.data());
+  common::OutputByteStream outStream(resultBuffer);
```
Can we use something from the standard library?

PHILO-HE replied:
Thanks for your comment! Not sure which standard library can be a replacement. And I just extended `OutputByteStream` to support appending multiple same char, required by outputting result inside `padContent()`. Maybe, we can keep the use of it.

mbasmanova commented on velox/functions/lib/DateTimeFormatter.cpp:
```diff
@@ -941,31 +1025,35 @@ std::string DateTimeFormatter::format(
       switch (token.pattern.specifier) {
         case DateTimeFormatSpecifier::ERA:
-          result += static_cast<signed>(calDate.year()) > 0 ? "AD" : "BC";
+          piece = static_cast<signed>(calDate.year()) > 0 ? "AD" : "BC";
```
Can we use constant values here and not copy strings?

mbasmanova commented on velox/functions/lib/DateTimeFormatter.cpp:
```diff
@@ -941,31 +1025,35 @@ std::string DateTimeFormatter::format(
           year = (year < 0 ? -year : year);
           auto century = year / 100;
-          result += padContent(century, '0', token.pattern.minRepresentDigits);
+          piece = padContent(century, '0', token.pattern.minRepresentDigits);
```
can we avoid creating intermediate strings?

PHILO-HE replied:
To also avoid intermediate strings inside `padContent()`, I just pass an `OutputByteStream` to `padContent` to directly write the content to result buffer in latest commit. Not sure whether it is a good implementation.

mbasmanova commented on velox/functions/lib/DateTimeFormatter.cpp:
```diff
@@ -189,28 +190,40 @@ constexpr int monthsFullLength[] = {7, 8, 5, 5, 3, 4, 4, 6, 9, 7, 8, 8};
-    char padding,
-    size_t totalDigits,
-    bool padFront = true) {
+    const char padding,
```
Drop const

mbasmanova commented on velox/functions/lib/DateTimeFormatter.cpp:
```diff
@@ -189,28 +190,40 @@ constexpr int monthsFullLength[] = {7, 8, 5, 5, 3, 4, 4, 6, 9, 7, 8, 8};
-    bool padFront = true) {
+    const char padding,
+    const size_t totalDigits,
+    common::OutputByteStream& outStream,
```
Why can't we just pass char* and use std::fill?

PHILO-HE replied:
Just changed to use std::fill/memcpy and remove the use of `OutputByteStream` in this patch. Thanks!

mbasmanova commented on velox/functions/lib/DateTimeFormatter.cpp:
```diff
@@ -189,28 +190,40 @@ constexpr int monthsFullLength[] = {7, 8, 5, 5, 3, 4, 4, 6, 9, 7, 8, 8};
   auto digitLength = contentLength - (isNegative ? 1 : 0);
   if (digitLength >= totalDigits) {
-    return strContent;
+    outStream.append(strContent.data(), strContent.size());
```
Can we eliminate intermediate string strContent?

This could use memcpy.

PHILO-HE replied:
It seems we have to use an intermediate string. As we need it to keep the string result of cast by folly: `const std::string strContent = folly::to<std::string>(content)`. Right?

mbasmanova replied:
@PHILO-HE Can we use std::to_chars instead of folly::to? https://en.cppreference.com/w/cpp/utility/to_chars

PHILO-HE commented on velox/expression/CastExpr-inl.h:
```diff
@@ -51,27 +52,6 @@ inline std::exception_ptr makeBadCastException(
 };
 
-// Copied from format.h of fmt.
-inline int countDigits(uint128_t n) {
```
Moved it into a new file named `IntegerUtil.h`.

pedroerp commented:
@PHILO-HE @mbasmanova I commented in the other discussion, but could we instead make the date formatter object take a complex type writer (sink/ostream or similar abstraction), since these things are in the end of the day already backed by a larger FlatVector? Then we can make sure that possible reallocations are amortized among different rows, or just by chaining the string buffers

PHILO-HE commented:
> @PHILO-HE @mbasmanova I commented in the other discussion, but could we instead make the date formatter object take a complex type writer (sink/ostream or similar abstraction), since these things are in the end of the day already backed by a larger FlatVector? Then we can make sure that possible reallocations are amortized among different rows, or just by chaining the string buffers

@pedroerp, thanks for your advice! Friendly ping @mbasmanova, what do you think? I'm not sure on this.

mbasmanova commented:
@PHILO-HE I'm not sure either. 

@pedroerp Do you feel strongly about this? Would you provide some guidance on how to proceed?

pedroerp commented:
@mbasmanova @PHILO-HE think it would be nice to provide a ostream-like API for returned strings in simple functions, but that might require some more thought. 

I'm ok with creating an Issue and merging this one, but I guess some of these changes might need to be cleaned up at some point.

mbasmanova commented:
> Hi @mbasmanova, do we still need this PR? Pedro's proposal sounds a more common fix.

I guess it depends on whether you'd like to implement Pedro's proposal instead.

PHILO-HE commented:
@mbasmanova @pedroerp, thanks for your time on this discussion!
I think we can close this one and try to work out a more common fix for such string buffer copy issues. @pedroerp, could you create an issue about your proposal?

PHILO-HE commented:
As discussed in other PR, this patch is still needed to avoid string copy before generic implementation is proposed. @mbasmanova, could you review further? Thanks!

mbasmanova commented on velox/type/IntegerUtil.h:
```diff
@@ -0,0 +1,44 @@
+
+// Copied from format.h of fmt.
+inline int countDigits(uint128_t n) {
+  int count = 1;
```
I'm not sure if velox/type is the right place for this helper function. Maybe velox/common/base/CountBits.h?

use FOLLY_ALWAYS_INLINE

mbasmanova commented on velox/type/IntegerUtil.h:
```diff
@@ -0,0 +1,44 @@
+
+#pragma once
+
+#include "velox/type/HugeInt.h"
```
Can we just use __uint128_t and not include this header?

mbasmanova commented on velox/functions/prestosql/DateTimeFunctions.h:
```diff
@@ -1132,46 +1132,48 @@ struct DateFormatFunction : public TimestampWithTimezoneSupport<T> {
-          std::string_view(formatString->data(), formatString->size()));
-      isConstFormat_ = true;
-    }
+  FOLLY_ALWAYS_INLINE void setFormatter(const arg_type<Varchar> formatString) {
```
Make this private and move after public methods.

mbasmanova replied:
const &

mbasmanova commented on velox/functions/prestosql/DateTimeFunctions.h:
```diff
@@ -1132,46 +1132,48 @@ struct DateFormatFunction : public TimestampWithTimezoneSupport<T> {
 
   const date::time_zone* sessionTimeZone_ = nullptr;
   std::shared_ptr<DateTimeFormatter> mysqlDateTime_;
+  uint32_t maxResultSize_;
```
Make these private and move after all methods.

mbasmanova replied:
@PHILO-HE Would you address these? Otherwise, looks great.

PHILO-HE replied:
@mbasmanova, I forgot this fix. Just updated in latest commit. Thanks!

mbasmanova commented on velox/functions/prestosql/DateTimeFunctions.h:
```diff
@@ -1132,46 +1132,48 @@ struct DateFormatFunction : public TimestampWithTimezoneSupport<T> {
-    if (resultSize != 0) {
-      std::memcpy(result.data(), formattedResult.data(), resultSize);
-    }
     return true;
```
If we always return true, then change function to void.

mbasmanova commented on velox/functions/lib/DateTimeFormatter.h:
```diff
@@ -175,10 +175,22 @@ class DateTimeFormatter {
 
+  /// Result buffer is pre-allocated according to maxResultSize.
+  /// Returns actual size.
+  int32_t format(
```
Is the other format method still used? Can we update callsites and remove it?

PHILO-HE replied:
Currently, this method is only used by DateTimeFormatterTest.cpp. Maybe, we can keep it here for possible use at other place.

mbasmanova replied:
We don't want this method to be used since it is slow. Let's remove it.

PHILO-HE replied:
Just removed it in latest commit. Thanks!

mbasmanova commented on velox/functions/lib/DateTimeFormatter.h:
```diff
@@ -175,10 +175,22 @@ class DateTimeFormatter {
 
+  /// Returns max size of the formatted string. Can be used to preallocate
+  /// memory before calling format() to avoid extra copy.
+  uint32_t maxResultSize(const date::time_zone* timezone) const;
```
Would be nice to add some unit tests.

PHILO-HE replied:
Just added some test cases. And the existing tests for DateTimeFormatter, format_datetime & date_format also cover the verification of the new API.

mbasmanova commented on velox/functions/lib/DateTimeFormatter.cpp:
```diff
@@ -180,37 +182,54 @@ static std::
 /// to be. If totalDigits is already smaller than content length, the original
-/// content will be returned with no padding
+/// content will be returned with no padding.
+/// \param result the pointer to string result.
```
use @param and document all parameters
use //

mbasmanova reviewed (commented):
Looks good overall % some comments. 

It would be nice to add a small benchmark to see how much improvement this brings. See velox/benchmarks/ExpressionBenchmarkBuilder.h

PHILO-HE commented:
@mbasmanova, this is the result of a small benchmark.

```text
============================================================================
[...]hmarks/ExpressionBenchmarkBuilder.cpp     relative  time/iter   iters/s
============================================================================
##format_datetime                                        337.01ms      2.97
----------------------------------------------------------------------------

============================================================================
[...]hmarks/ExpressionBenchmarkBuilder.cpp     relative  time/iter   iters/s
============================================================================
##format_datetime (remove extra copy, this PR)           167.07ms      5.99
----------------------------------------------------------------------------
----------------------------------------------------------------------------
```

mbasmanova commented:
@PHILO-HE Thank you for sharing benchmark results. Are you seeing 2x win? Let's update PR description and include the benchmark in this PR.

PHILO-HE commented:
> @PHILO-HE Thank you for sharing benchmark results. Are you seeing 2x win? Let's update PR description and include the benchmark in this PR.

Yes, it shows ~2x win. Just included this simple benchmark in this PR and updated PR description.

mbasmanova commented on velox/benchmarks/basic/FormatDateTimeBenchmark.cpp:
```diff
@@ -0,0 +1,49 @@
+      });
+
+  benchmarkBuilder
+      .addBenchmarkSet("", vectorMaker.rowVector({"input"}, {timestampInput}))
```
Do we need custom input or can we let the builder generate input using fuzzer?

mbasmanova commented on velox/benchmarks/basic/FormatDateTimeBenchmark.cpp:
```diff
@@ -0,0 +1,49 @@
+      .addExpression(
+          "format_datetime",
+          "format_datetime(input, 'yyyy-MM-dd HH:mm:ss.SSS')")
+      .withIterations(1000)
```
Is this needed? CC: @laithsakka

PHILO-HE replied:
1000 iterations may be better.

mbasmanova replied:
Curious, how do you decide whether 1000 is better than default (100?). Would you clarify? Perhaps, add a comment in the code to help future readers.

PHILO-HE replied:
@mbasmanova , I just found the default one is 1000. Maybe, we can just remove the iteration setting. Right?
https://github.com/facebookincubator/velox/blob/7822fe154b9d24d4e50bbc41f0c4ca5d39e2db03/velox/benchmarks/ExpressionBenchmarkBuilder.h#L91

mbasmanova replied:
Yes, let's remove. Thanks.

mbasmanova commented on velox/benchmarks/basic/FormatDateTimeBenchmark.cpp:
```diff
@@ -0,0 +1,49 @@
+  benchmarkBuilder
+      .addBenchmarkSet("", vectorMaker.rowVector({"input"}, {timestampInput}))
+      .addExpression(
+          "format_datetime",
```
Is this a valid expression? Don't we need parentheses after function name?

PHILO-HE replied:
Hi @mbasmanova, this is just a name for this expression test. Maybe, no need to name it. Just removed it. Only keep the benchmark test name passed to `addBenchmarkSet`.

pedroerp commented:
> @mbasmanova @pedroerp, thanks for your time on this discussion! I think we can close this one and try to work out a more common fix for such string buffer copy issues. @pedroerp, could you create an issue about your proposal?

@PHILO-HE Created an Issue to track it: https://github.com/facebookincubator/velox/issues/8035

PHILO-HE commented:
Hi @mbasmanova, do you have any other comment? Thanks!

PHILO-HE commented:
Hi @mbasmanova, I have addressed all comments. The CI failure is a known issue: https://github.com/facebookincubator/velox/issues/7954, not related to this PR.

mbasmanova reviewed (approved):
Thanks. Please, rebase so I can merge.

PHILO-HE commented:
> Thanks. Please, rebase so I can merge.

@mbasmanova, already rebased and squashed into one commit. Thanks for your time!
